// @sanity/runtime-cli/dist/utils/functions/transpile/function.js
import { mkdir, readFile, rm, stat, writeFile } from 'node:fs/promises';
import path from 'node:path';
import { performance } from 'node:perf_hooks';
import { cwd } from 'node:process';
import * as find from 'empathic/find';
import { build as viteBuild } from 'vite';
import { detectNativeModulesPlugin } from '../detect-native-modules.js';
import { findFunctionEntryPoint } from '../find-entry-point.js';
import { cleanupSourceMaps } from './cleanup-source-maps.js';
import { createPnpmRequire } from './pnpm.js';
import { verifyHandler } from './verify-handler.js';
export const resolveViteRoot = (entry) => {
    const nodeModulesDir = find.dir('node_modules', { cwd: entry });
    return nodeModulesDir ? path.dirname(nodeModulesDir) : entry;
};
export async function transpileFunction(resource, { installer } = {}) {
    if (!resource.src)
        throw new Error('Resource src is required');
    if (!resource.name)
        throw new Error('Resource name is required');
    const timings = {};
    const transpileStart = performance.now();
    const sourcePath = path.resolve(cwd(), resource.src);
    const stats = await stat(sourcePath);
    const fnDisplayName = resource.displayName ?? resource.name;
    const findEntryStart = performance.now();
    const entry = await findFunctionEntryPoint(sourcePath, fnDisplayName);
    timings['transpile:findEntry'] = performance.now() - findEntryStart;
    const entryDir = stats.isFile() ? path.dirname(sourcePath) : sourcePath;
    const outputPathStart = performance.now();
    const outputDir = await getTranspileOutputPath(entryDir, resource.name);
    const outputFile = path.join(outputDir, getOutputFilename(entry));
    const fnRootDir = resolveViteRoot(entryDir);
    timings['transpile:setupOutput'] = performance.now() - outputPathStart;
    async function cleanupTmpDir() {
        // Feel a certain way about leaving things uncleaned, but helps with debugging for now
        // await rm(outputDir, {recursive: true, force: true}).catch(logCleanupFailure)
    }
    try {
        const viteStart = performance.now();
        // Always bundle the code if pnpm usage is detected or specified by the user
        const bundle = !!find.file('pnpm-workspace.yaml') || installer === 'pnpm';
        const result = await viteBuild({
            root: fnRootDir,
            logLevel: 'silent',
            build: {
                target: 'node20',
                outDir: outputDir,
                emptyOutDir: false,
                minify: false,
                sourcemap: true,
                ssr: true,
                rolldownOptions: {
                    input: entry,
                    output: {
                        format: 'esm',
                        entryFileNames: getOutputFilename(entry),
                        ...(bundle
                            ? {}
                            : {
                                preserveModules: true,
                                preserveModulesRoot: fnRootDir,
                            }),
                    },
                    external: bundle ? [] : [/node_modules/],
                },
            },
            resolve: {
                // Replaces the vite-tsconfig-paths plugin; vite 8 resolves tsconfig
                // `compilerOptions.paths` natively
                tsconfigPaths: true,
            },
            ssr: {
                noExternal: bundle ? true : [],
                resolve: {
                    conditions: ['node'],
                },
            },
            plugins: bundle ? [detectNativeModulesPlugin(), pnpmResolvePlugin(cwd())] : [],
        });
        timings['transpile:build'] = performance.now() - viteStart;
        const verifyStart = performance.now();
        await verifyHandler(result);
        timings['transpile:verify'] = performance.now() - verifyStart;
        const pkgStart = performance.now();
        await writeTranspiledPackageJson(entryDir, outputFile);
        timings['transpile:writePackage'] = performance.now() - pkgStart;
        const cleanupStart = performance.now();
        await cleanupSourceMaps(sourcePath, outputDir);
        timings['transpile:cleanupMaps'] = performance.now() - cleanupStart;
        timings.bundle = performance.now() - transpileStart;
        return {
            type: 'success',
            outputDir,
            warnings: [],
            cleanup: cleanupTmpDir,
            timings,
            bundled: bundle,
        };
    }
    catch (err) {
        await cleanupTmpDir();
        throw new Error(`Transpiling of function failed: ${formatBuildError(err)}`, {
            cause: err,
        });
    }
}
/**
 * Vite 8's `build()` wraps failures in a `BundleError` whose `message` contains
 * a "Build failed with N errors" preamble and full stack traces. The individual
 * errors on `.errors` carry the original, human-readable messages.
 */
export function formatBuildError(err) {
    if (!(err instanceof Error))
        return String(err);
    if ('errors' in err && Array.isArray(err.errors) && err.errors.length > 0) {
        return err.errors.map(formatBuildErrorEntry).join('\n');
    }
    return err.message;
}
/**
 * A `BundleError.errors` entry may be...
 *   a real `Error`,
 *   a plain object carrying a `message` (Rolldown/Oxc diagnostics),
 *   or unknown
 * Prefer `.message` and fall back to `String()`
 */
function formatBuildErrorEntry(e) {
    if (e instanceof Error)
        return e.message;
    const message = e?.message;
    return typeof message === 'string' ? message : String(e);
}
async function writeTranspiledPackageJson(inputDir, outputFilePath) {
    const baseName = path.basename(outputFilePath);
    let original;
    try {
        const pkgJsonPath = path.join(inputDir, 'package.json');
        original = JSON.parse(await readFile(pkgJsonPath, 'utf-8'));
    }
    catch {
        original = undefined;
    }
    const bundled = {
        // One could argue that we should strip this down significantly.
        // Theoretically though, a function may reach into it to get dependency versions
        // and whatnot, so maybe it makes more sense to keep it as close to the original
        // as possible?
        ...original,
        main: baseName, // This should never be the input file, always the built output name
        type: 'module', // We explicitly create ESM output
    };
    const pkgJsonOutputPath = path.join(path.dirname(outputFilePath), 'package.json');
    await writeFile(pkgJsonOutputPath, JSON.stringify(bundled, null, 2));
}
async function getTranspileOutputPath(entryDir, fnName) {
    const tmpPath = path.resolve(entryDir, '.build', `function-${fnName}`);
    await rm(tmpPath, { recursive: true, force: true }).catch(logCleanupFailure);
    await mkdir(tmpPath, { recursive: true });
    return tmpPath;
}
/**
 * Minor convenience/niceness to keep the same input filename, but change the extension
 */
function getOutputFilename(entryFileName) {
    const baseName = path.basename(entryFileName, path.extname(entryFileName));
    return baseName ? `${baseName}.js` : 'index.js';
}
function logCleanupFailure(err) {
    console.warn(`[warn] Failed to clean up temporary files: ${err instanceof Error ? err.message : err}`);
}
/**
 * Vite plugin to resolve packages from pnpm's virtual store.
 * We don't have a node_modules dir when we bundle with pnpm, so we need a plugin to resolve dependencies from pnpm's virtual store
 * In pnpm workspaces, hoisted dependencies live in node_modules/.pnpm/node_modules/
 * which isn't reachable via standard Node resolution.
 */
function pnpmResolvePlugin(workspaceRoot) {
    const pnpmRequire = createPnpmRequire(workspaceRoot);
    return {
        name: 'pnpm-resolve',
        async resolveId(source, importer, options) {
            // Only act as fallback — let Vite try first
            const resolved = await this.resolve(source, importer, {
                ...options,
                skipSelf: true,
            });
            if (resolved)
                return resolved;
            // Fallback: use Node's require.resolve() from the pnpm virtual store.
            // This properly follows package.json exports/main to the entry file.
            try {
                return pnpmRequire.resolve(source);
            }
            catch {
                return null;
            }
        },
    };
}
