// @sanity/runtime-cli/dist/cores/blueprints/deploy.js
import { stashAsset } from '../../actions/blueprints/assets.js';
import { getStack, updateStack } from '../../actions/blueprints/stacks.js';
import { waitForOperation } from '../../actions/blueprints/wait-for-operation.js';
import { checkUserPermission } from '../../actions/sanity/access.js';
import { createHintCollector } from '../../utils/blueprints/hints.js';
import { niceId } from '../../utils/display/presenters.js';
import { CODE_OPERATION_UNCONFIRMED, EXIT_OPERATION_UNCONFIRMED } from '../../utils/exit-codes.js';
import { styleText } from '../../utils/style-text.js';
import { isAssetResource } from '../../utils/types.js';
const DEFAULT_ASSET_TIMEOUT = 60;
const assetTimeoutS = Number(process.env.SANITY_ASSET_TIMEOUT) || DEFAULT_ASSET_TIMEOUT;
const assetTimeoutMs = assetTimeoutS * 1000;
const warnTimeoutMs = assetTimeoutMs / 2;
export async function blueprintDeployCore(options) {
    const { bin = 'sanity', log, auth, stackId, scopeType, scopeId, deployedStack, blueprint, flags, } = options;
    const { verbose } = flags;
    const noWait = flags['no-wait'] || false;
    const stackName = flags['new-stack-name'] ?? deployedStack.name;
    const installer = flags['fn-installer'] || undefined;
    log(`Deploying "${stackName}" ${niceId(deployedStack.id)}...`);
    try {
        const { ok: checkOk, stack: currentStack } = await getStack({ stackId, auth, logger: log });
        if (!checkOk) {
            return { success: false, error: 'Failed to check current Stack status' };
        }
        const operationStatus = currentStack.recentOperation?.status;
        if (operationStatus === 'QUEUED' || operationStatus === 'IN_PROGRESS') {
            return {
                success: false,
                error: 'A Stack operation is already in progress.',
                suggestions: [
                    `Run \`npx ${bin} blueprints info\` to check operation status.`,
                    'Wait for the current operation to complete, then try again.',
                ],
            };
        }
        const resources = [...blueprint.resources];
        const assetResources = resources.filter(isAssetResource);
        if (assetResources.length > 0) {
            log('Processing assets...');
            for (const resource of assetResources) {
                const preDeployResult = await preDeploy(resource, { auth, installer, log });
                if (!preDeployResult.success)
                    return preDeployResult;
                const idx = resources.indexOf(resource);
                if (idx !== -1)
                    resources[idx] = preDeployResult.resource;
            }
        }
        const spinner = log.ora('Deploying...').start();
        const { ok: deployOk, stack, error: deployError, } = await updateStack({
            stackId,
            stackMutation: {
                scopeType,
                scopeId,
                name: stackName,
                document: { resources },
                userMessage: flags.message,
            },
            auth,
            logger: log,
        });
        if (!deployOk) {
            spinner.fail(`${styleText('red', 'Failed')} to update Stack deployment`);
            const hints = createHintCollector(bin);
            if (deployError) {
                hints.inspectMessage(deployError);
            }
            return {
                success: false,
                error: deployError || 'Failed to update Stack deployment',
                suggestions: [
                    ...hints.getSuggestions(),
                    `Run \`npx ${bin} blueprints plan\` to preview changes before deploying.`,
                    `Run \`npx ${bin} blueprints doctor\` to check your configuration.`,
                ],
            };
        }
        spinner.stop().clear();
        const legacyPermissionsNoticePromise = CHECK_LEGACY_DEPLOY_PERMISSION
            ? checkLegacyDeployPermission({
                auth,
                scopeType,
                scopeId,
                log,
            })
            : Promise.resolve(null);
        if (noWait) {
            log(styleText(['bold', 'green'], 'Stack deployment started!'));
            log(`Use \`npx ${bin} blueprints info\` to check status`);
            const legacyWarning = await legacyPermissionsNoticePromise;
            return {
                success: true,
                json: { stackId, resources },
                data: { resources },
                warnings: legacyWarning ? [legacyWarning] : undefined,
            };
        }
        log(styleText('dim', 'Stack deployment progress:'));
        log('');
        const outcome = await waitForOperation({
            stackId,
            operationId: stack.operationId,
            auth,
            log,
            bin,
            verbose,
            progressNoun: 'deployment',
        });
        if (outcome.type === 'completed') {
            log(styleText(['bold', 'green'], 'Stack deployment completed!'));
            const legacyWarning = await legacyPermissionsNoticePromise;
            return {
                success: true,
                json: { stackId, resources },
                data: { resources },
                warnings: legacyWarning ? [legacyWarning] : undefined,
            };
        }
        if (outcome.type === 'unconfirmed') {
            return {
                success: false,
                error: `Stack deployment was accepted but completion could not be confirmed${outcome.error ? ` (${outcome.error})` : ''}.`,
                code: CODE_OPERATION_UNCONFIRMED,
                exitCode: EXIT_OPERATION_UNCONFIRMED,
                suggestions: [
                    'The deployment may or may not have finished on Sanity servers.',
                    `Run \`npx ${bin} blueprints info\` to check status.`,
                    `Run \`npx ${bin} blueprints logs --watch\` to keep streaming logs.`,
                ],
            };
        }
        return {
            success: false,
            error: 'Stack deployment failed',
            suggestions: [
                ...outcome.logHints,
                'Review the deployment output above for error details.',
                `Run \`npx ${bin} blueprints logs --verbose\` for more context.`,
                `Run \`npx ${bin} blueprints plan\` to identify issues with your Blueprint.`,
            ],
        };
    }
    catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        return { success: false, error: errorMessage };
    }
}
async function preDeploy(resource, options) {
    const { auth, installer, log } = options;
    const fnSpinner = log.ora({ text: `Processing ${resource.name}...`, prefixText: ' ' }).start();
    const warnTimer = setTimeout(() => {
        fnSpinner.text = `Still processing ${resource.name}, this can take a moment...`;
    }, warnTimeoutMs);
    let assetTimeoutTimer;
    let result;
    try {
        result = await Promise.race([
            stashAsset({ resource, auth, logger: log, installer }),
            new Promise((_, reject) => {
                assetTimeoutTimer = setTimeout(() => {
                    reject(new Error(`Processing ${resource.name} timed out after ${assetTimeoutS}s`));
                }, assetTimeoutMs);
            }),
        ]);
    }
    catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        fnSpinner.fail(msg);
        return { success: false, error: msg };
    }
    finally {
        clearTimeout(warnTimer);
        clearTimeout(assetTimeoutTimer);
    }
    if (result.success && result.assetId) {
        const updatedResource = { ...resource, src: result.assetId };
        fnSpinner.succeed(`${updatedResource.name} ${niceId(result.assetId)}`);
        log(`   Source: ${updatedResource.src}`);
        if (result.hash) {
            if (result.hash.length > 24) {
                log.verbose(`   Hash: ${result.hash.slice(0, 8)}...${result.hash.slice(-12)}`);
            }
            else {
                log.verbose(`   Hash: ${result.hash}`);
            }
        }
        if (result.exists)
            log.verbose('   Asset unchanged');
        return { success: true, resource: updatedResource };
    }
    const errorMsg = `Failed uploading ${resource.name} asset, deploy has stopped`;
    fnSpinner.fail(errorMsg);
    return { success: false, error: result.error || 'Failed to process asset' };
}
const CHECK_LEGACY_DEPLOY_PERMISSION = false;
const BLUEPRINTS_PERMISSION_ENFORCEMENT_DATE = 'July 1, 2026';
async function checkLegacyDeployPermission({ auth, scopeType, scopeId, log, }) {
    const requiredPermission = scopeType === 'organization' ? 'sanity.blueprints.deploy' : 'sanity.project.blueprints.deploy';
    try {
        const result = await checkUserPermission({
            auth,
            resourceType: scopeType,
            resourceId: scopeId,
            permission: requiredPermission,
            logger: log,
        });
        if (!result.ok) {
            log.verbose(`Failed to check ${requiredPermission}: ${result.error ?? 'unknown error'}`);
            return null;
        }
        if (result.hasPermission)
            return null;
        return {
            label: 'Legacy deploy permission',
            suggestions: [
                `This deploy used a legacy permission. Starting ${BLUEPRINTS_PERMISSION_ENFORCEMENT_DATE}, deploys will require the "${requiredPermission}" permission.`,
                scopeType === 'project'
                    ? 'Ask a project administrator to add you to a role with Blueprint deploy access: Administrator, Developer, or Blueprints Deployer. For CI, use a Blueprints Deployer API token.'
                    : 'Ask an organization administrator to grant you the Administrator role. For CI, use a Blueprints Deployer API token created by someone with access to this organization.',
            ],
        };
    }
    catch (err) {
        log.verbose(`Failed to check ${requiredPermission}: ${err instanceof Error ? err.message : String(err)}`);
        return null;
    }
}
