// @portabletext/editor/lib/_chunks-es/get-first-child.js
import { getAncestors, isObject as isObject$1, isEditableContainer, getNodes, comparePoints$1 as comparePoints, rangeEdges, isAfterPoint, comparePoints as comparePoints$1, comparePaths } from "./get-path-sub-schema.js";
import { getChildren, getNode, isKeyedSegment } from "./get-parent.js";
import { isSpan, isTextBlock, getSubSchema } from "@portabletext/schema";
function getAncestor(snapshot, path, options) {
  const {
    match,
    mode = "lowest"
  } = options, ancestors = getAncestors(snapshot, path);
  if (mode === "highest") {
    for (const ancestor of [...ancestors].reverse())
      if (match(ancestor.node, ancestor.path))
        return ancestor;
    return;
  }
  for (const ancestor of ancestors)
    if (match(ancestor.node, ancestor.path))
      return ancestor;
}
function isLeafObject(snapshot, node, path) {
  return isObject$1(snapshot, node) && !isEditableContainer(snapshot, node, path);
}
function getLeaf(snapshot, path, options) {
  const {
    edge
  } = options;
  let currentPath = path;
  if (currentPath.length === 0) {
    const children = getChildren(snapshot, []);
    if (children.length === 0)
      return;
    const firstOrLast = edge === "end" ? children.at(-1) : children.at(0);
    if (getChildren(snapshot, firstOrLast.path).length === 0)
      return firstOrLast;
    currentPath = firstOrLast.path;
  } else {
    const entry = getNode(snapshot, currentPath);
    if (!entry)
      return;
    if (getChildren(snapshot, currentPath).length === 0)
      return entry;
  }
  for (; ; ) {
    const children = getChildren(snapshot, currentPath);
    if (children.length === 0)
      return getNode(snapshot, currentPath) ?? void 0;
    currentPath = (edge === "end" ? children.at(-1) : children.at(0)).path;
  }
}
function isPath(value) {
  return Array.isArray(value) && (value.length === 0 || typeof value[0] == "number" || typeof value[0] == "string" || isKeyedSegment(value[0]));
}
const isObject = (value) => typeof value == "object" && value !== null;
function isPoint(value) {
  return isObject(value) && typeof value.offset == "number" && isPath(value.path);
}
function isRange(value) {
  return isObject(value) && isPoint(value.anchor) && isPoint(value.focus);
}
function getText(snapshot, path) {
  const entry = getNode(snapshot, path);
  if (!entry)
    return;
  if (isSpan({
    schema: snapshot.context.schema
  }, entry.node))
    return entry.node.text;
  let text = "";
  for (const descendant of getNodes(snapshot, {
    at: path
  }))
    isSpan({
      schema: snapshot.context.schema
    }, descendant.node) && (text += descendant.node.text);
  return text;
}
function getSpan(snapshot, path) {
  const entry = getNode(snapshot, path);
  if (entry && isSpan({
    schema: snapshot.context.schema
  }, entry.node))
    return {
      node: entry.node,
      path: entry.path
    };
}
function pathContains(ancestor, descendant) {
  if (ancestor.length > descendant.length)
    return !1;
  for (let i = 0; i < ancestor.length; i++) {
    const segment = ancestor[i], otherSegment = descendant[i];
    if (isKeyedSegment(segment) && isKeyedSegment(otherSegment)) {
      if (segment._key !== otherSegment._key)
        return !1;
    } else if (segment !== otherSegment)
      return !1;
  }
  return !0;
}
function isBeforePoint(point, another, root) {
  return comparePoints(point, another, root) === -1;
}
function rangeIntersects(snapshot, range, target) {
  if (!range || target === null)
    return !1;
  const root = {
    value: snapshot.context.value
  };
  if (isRange(target)) {
    if (rangeIntersects(snapshot, range, target.anchor) || rangeIntersects(snapshot, range, target.focus))
      return !0;
    const [rs, re] = rangeEdges(range, root), [ts, te] = rangeEdges(target, root);
    return isBeforePoint(rs, ts, root) && isAfterPoint(re, te, root);
  }
  const [start, end] = rangeEdges(range, root);
  let isAfterStart = !1, isBeforeEnd = !1;
  return isPoint(target) ? (isAfterStart = comparePoints$1(snapshot, target, start) >= 0, isBeforeEnd = comparePoints$1(snapshot, target, end) <= 0) : (isAfterStart = comparePaths(target, start.path, root) >= 0, isBeforeEnd = comparePaths(target, end.path, root) <= 0), isAfterStart && isBeforeEnd;
}
function getTextBlock(snapshot, path) {
  const entry = getNode(snapshot, path);
  if (entry && isTextBlock({
    schema: snapshot.context.schema
  }, entry.node))
    return {
      node: entry.node,
      path: entry.path
    };
}
function getUnionSchema(schema, containers) {
  const decorators = mergeByName(schema.decorators, []), annotations = mergeByName(schema.annotations, []), lists = mergeByName(schema.lists, []), styles = mergeByName(schema.styles, []), inlineObjects = mergeByName(schema.inlineObjects, []), blockObjects = mergeByName(schema.blockObjects, []);
  for (const container of containers.values()) {
    if (!acceptsTextBlock(container.field.of, schema.block.name))
      continue;
    const sub = getSubSchema(schema, container.field.of);
    mergeByName(sub.decorators, decorators), mergeByName(sub.annotations, annotations), mergeByName(sub.lists, lists), mergeByName(sub.styles, styles), mergeByName(sub.inlineObjects, inlineObjects), mergeByName(sub.blockObjects, blockObjects);
  }
  return {
    ...schema,
    decorators,
    annotations,
    lists,
    styles,
    inlineObjects,
    blockObjects
  };
}
function acceptsTextBlock(of, blockName) {
  return of.some((member) => member.type === "block" || member.type === blockName);
}
function mergeByName(source, target) {
  for (const entry of source)
    target.some((existing) => existing.name === entry.name) || target.push(entry);
  return target;
}
function getFirstChild(snapshot, path) {
  return getChildren(snapshot, path).at(0);
}
export {
  getAncestor,
  getFirstChild,
  getLeaf,
  getSpan,
  getText,
  getTextBlock,
  getUnionSchema,
  isBeforePoint,
  isLeafObject,
  isObject,
  isPath,
  isPoint,
  isRange,
  pathContains,
  rangeIntersects
};
